src/core/qgsmaphittest.cpp

```diff
@@ -203,3 +257,143 @@ void QgsMapHitTest::runHitTestLayer( QgsVectorLayer *vl, SymbolSet &usedSymbols,
+  : QgsTask( tr( "Updating Legend" ), QgsTask::Flag::CanCancel | QgsTask::Flag::CancelWithoutPrompt | QgsTask::Flag::Silent )
+  , mSettings( settings )
+  , mLayerFilterExpression( layerFilterExpression )
+  , mPolygon( polygon )
+  , mOnlyExpressions( false )
+{
+  prepare();
+}
+
+QgsMapHitTestTask::QgsMapHitTestTask( const QgsMapSettings &settings, const QgsMapHitTest::LayerFilterExpression &layerFilterExpression )
+  : QgsTask( tr( "Updating Legend" ), QgsTask::Flag::CanCancel | QgsTask::Flag::CancelWithoutPrompt | QgsTask::Flag::Silent )
+  , mSettings( settings )
+  , mLayerFilterExpression( layerFilterExpression )
+  , mOnlyExpressions( true )
+{
+  prepare();
+}
+
+QMap<QString, QSet<QString> > QgsMapHitTestTask::results() const
+{
+  return mResults;
+}
+
+///@cond PRIVATE
+QMap<QString, QList<QString> > QgsMapHitTestTask::resultsPy() const
+{
+  QMap<QString, QList<QString> > res;
+  for ( auto it = mResults.begin(); it != mResults.end(); ++it )
+  {
+    res.insert( it.key(), qgis::setToList( it.value() ) );
+  }
+  return res;
+}
+///@endcond PRIVATE
+
+void QgsMapHitTestTask::prepare()
+{
+  const QList< QgsMapLayer * > layers = mSettings.layers( true );
+  for ( QgsMapLayer *layer : layers )
+  {
+    QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( layer );
+    if ( !vl || !vl->renderer() )
+      continue;
+
+    QgsMapLayerStyleOverride styleOverride( vl );
+    if ( mSettings.layerStyleOverrides().contains( vl->id() ) )
+      styleOverride.setOverrideStyle( mSettings.layerStyleOverrides().value( vl->id() ) );
+
+    if ( !mOnlyExpressions )
+    {
+      if ( !vl->isInScaleRange( mSettings.scale() ) )
+      {
+        continue;
+      }
+    }
+
+    PreparedLayerData layerData;
+    layerData.source = std::make_unique< QgsVectorLayerFeatureSource >( vl );
+    layerData.layerId = vl->id();
+    layerData.crs = vl->crs();
+    layerData.fields = vl->fields();
+    layerData.renderer.reset( vl->renderer()->clone() );
+    layerData.transform = mSettings.layerTransform( vl );
+    layerData.extent = mSettings.outputExtentToLayerExtent( vl, mSettings.visibleExtent() );
+    layerData.layerScope.reset( QgsExpressionContextUtils::layerScope( vl ) );
+
+    mPreparedData.emplace_back( std::move( layerData ) );
+  }
+}
+
+void QgsMapHitTestTask::cancel()
+{
+  if ( mFeedback )
+    mFeedback->cancel();
+
+  QgsTask::cancel();
+}
+
+bool QgsMapHitTestTask::run()
+{
+  mFeedback = std::make_unique< QgsFeedback >();
+  connect( mFeedback.get(), &QgsFeedback::progressChanged, this, &QgsTask::progressChanged );
+
+  std::unique_ptr< QgsMapHitTest > hitTest;
+  if ( !mOnlyExpressions )
+    hitTest = std::make_unique< QgsMapHitTest >( mSettings, mPolygon, mLayerFilterExpression );
+  else
+    hitTest = std::make_unique< QgsMapHitTest >( mSettings, mLayerFilterExpression );
+
+  // TODO: do we need this temp image?
```

Comment:
Has that been explored yet?

Reply:
It's a long-standing TODO from the non-threaded code. I believe it's there to be super-safe, just in case some particular symbol layer type or renderer assumes that a paint device will always be available. I'm reticent to remove it.